### PR TITLE
fix(atomic): added focus trap to refine modal

### PR DESCRIPTION
**Pull request by btaillon**
https://coveord.atlassian.net/browse/KIT-1085

This is pretty flaky on iOS, because Voiceover for iOS refuses to focus on the close button inside of the modal. Sometimes, it won't be possible to move the focus until the user forces the focus at any other random place on the modal.

Deactivating the ShadowDOM seems to fix the issue, but I'm not sure if this is something we should do, and if so, I would do that in a separate PR. What do you think?

### CODE REVIEW

File: packages/atomic/src/components/common/atomic-focus-trap.tsx [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  if (element.parentNode) {
+    return element.parentNode as Element | ShadowRoot;
+  }
+  if (element instanceof ShadowRoot) {
+    return element.host;
+  }
+  return null;
+}
+
+function contains(
```
Comment (btaillon) [inline packages/atomic/src/components/common/atomic-focus-trap.tsx:13]: These two utils could be extracted to a separate file, if you prefer.

File: packages/atomic/src/components/common/atomic-focus-trap.tsx [hunk 2]
```diff
@@ -0,0 +1,112 @@
+import {Component, Element, Listen, Prop, Watch} from '@stencil/core';
```
Comment (btaillon) [inline packages/atomic/src/components/common/atomic-focus-trap.tsx:1]: I wasn't sure in which folder to put this component, since the other components in `common` are functional components. Do you think that's the correct place?

File: packages/atomic/src/components/common/atomic-focus-trap.tsx [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  }
+
+  showAll() {
+    let el: Element | undefined;
+    while ((el = this.hiddenElements.pop())) {
+      el.removeAttribute('aria-hidden');
+    }
+  }
+
+  isolate(element: Element | ShadowRoot) {
```
Comment (samisayegh) [inline packages/atomic/src/components/common/atomic-focus-trap.tsx:73]: Could `hideSiblings` be a clearer name for this method?

[the remark this answers is not in the document]
Comment (samisayegh) [reply packages/atomic/src/components/common/atomic-focus-trap.tsx:73]: Personally, I needed to read the function implementation to understand what it was doing,

File: packages/atomic/src/components/atomic-refine-modal/atomic-refine-modal.tsx [hunk 4]
```diff
@@ -73,14 +77,27 @@ export class AtomicRefineModal implements InitializableComponent {
...
   }
 
+  @DeferUntilRender()
+  private focusOnCloseButton() {
+    this.closeButton?.focus();
+  }
+
+  @DeferUntilRender()
+  private focusOnOpenButton() {
+    console.log(this.openButton);
```
[the remark this answers is not in the document]
Comment (btaillon) [reply packages/atomic/src/components/atomic-refine-modal/atomic-refine-modal.tsx:97]: Woops, yes.

File: packages/atomic/src/components/common/atomic-focus-trap.tsx (hunk 3 above)
Comment (btaillon) [reply packages/atomic/src/components/common/atomic-focus-trap.tsx:73]: Maybe. However, I would expect `hideSiblings` not to be recursive, so maybe `hideSiblingsRecursively`?

**Comment by ThibodeauJF** [comment]
Let's keep the shadow dom 😄 If we go down that path we might end up removing it in too many places

File: packages/atomic/src/components/common/atomic-focus-trap.tsx [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  return element.querySelector(
+    focusableSelectors.join(',')
+  ) as HTMLElement | null;
+}
+
+@Component({
+  tag: 'atomic-focus-trap',
+  shadow: false,
+})
+export class AtomicFocusTrap {
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/common/atomic-focus-trap.tsx:53]: Is this an internal component?

File: packages/atomic/src/components/atomic-refine-modal/atomic-refine-modal.tsx [hunk 6]
```diff
@@ -73,14 +77,27 @@ export class AtomicRefineModal implements InitializableComponent {
...
         break;
       case 'closed':
         document.body.classList.remove(modalOpenedClass);
         this.flushFacetElements();
+        this.focusOnOpenButton();
         break;
     }
   }
 
+  @DeferUntilRender()
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/atomic-refine-modal/atomic-refine-modal.tsx:90]: why was this needed?

File: packages/atomic/src/components/common/atomic-focus-trap.tsx (hunk 3 above)
Comment (ThibodeauJF) [reply packages/atomic/src/components/common/atomic-focus-trap.tsx:73]: `hideFamily` 😂

File: packages/atomic/src/components/common/atomic-focus-trap.tsx (hunk 5 above)
Comment (btaillon) [reply packages/atomic/src/components/common/atomic-focus-trap.tsx:53]: Yes, mostly. I forgot `@internal`.

File: packages/atomic/src/components/atomic-refine-modal/atomic-refine-modal.tsx (hunk 6 above)
Comment (btaillon) [reply packages/atomic/src/components/atomic-refine-modal/atomic-refine-modal.tsx:90]: The button doesn't exist yet

File: packages/atomic/src/components/common/atomic-focus-trap.tsx (hunk 2 above)
Comment (olamothe) [reply packages/atomic/src/components/common/atomic-focus-trap.tsx:1]: I mean... I don't remember having ever worked on a software project that did not have that kind of "put your code here if you are not sure" type of folder 😂.

So 🤷‍♂️, yeah it's fine IMO.

File: packages/atomic/src/components/common/atomic-focus-trap.tsx (hunk 1 above)
Comment (olamothe) [reply packages/atomic/src/components/common/atomic-focus-trap.tsx:13]: It's better to keep it internal if it's not used anywhere else.

And separate/extract only when it's needed.

File: packages/atomic/src/components/common/atomic-focus-trap.tsx (hunk 3 above)
Comment (olamothe) [reply packages/atomic/src/components/common/atomic-focus-trap.tsx:73]: `hideSiblingsAndParents` ?

**Review by olamothe — APPROVED** [review]

File: packages/atomic/src/components/common/atomic-focus-trap.tsx (hunk 3 above)
Comment (btaillon) [reply packages/atomic/src/components/common/atomic-focus-trap.tsx:73]: I'll use `hideSiblingsRecursively`, because technically every ancestor must be visible.
